keps/sig-node/4580-deprecate-kubelet-runonce/README.md

```diff
@@ -0,0 +1,348 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Deprecated and removed kubelet support for RunOnce mode.
+
+## Motivation
+
+* Runonce does not work in systemd mode.
+* Runonce mode doesn't support many newer pod features (init containers)
+* The pod lifecycle for runonce mode is even less well defined than normal.
+* The mode appears to be entirely untested in CI, except for a few unit tests.
+
+### Goals
+
+* Mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated.
+* Remove kubelet support for RunOnce mode.
+
+### Non-Goals
+
+Immediate removal: the deprecation and removal process will be gradual and feature gate to increase awareness among potential users.
+
+## Proposal
+
+The runonce mode of kubelet will exit the kubelet process after spawning pods from the local manifests or remote URL. It is suitable for scenarios where one-time tasks need to be run on the node, this proposal outlines plans to deprecate and remove runonce mode in kubelet.
+
+### Risks and Mitigations
+
+Some people may still rely on this feature, but podman addresses the same use case with more well-supported way, ref: https://docs.podman.io/en/latest/markdown/podman-kube.1.html. Affected users can migrate to *podman kube subcommand* on demand.
+
+## Design Details
+
+### KubeletConfiguration Change: KubeletConfiguration
+
+Mark the `RunOnce` field as Deprecated.
+
+### kubelet flag Change
+
+make the `runonce` flag as Deprecated.
+
+### Implement warning logging for runonce mode usage
+
+Starting in 1.31, during kubelet startup, if running in runonce mode, the kubelet will log a warning message, for example:
+
+```
+klog.Warning("RunOnce mode has been deprecated, and will be removed in a future release")
+```
+
+### Introduction RemoveRunonce feature gate
+
+With the introduction of the `RemoveRunonce` feature gate, Kubernetes aims to guide users through the deprecated runonce mode. When this feature gate is enabled, kubelet will refuse to start when the runonce flag is set.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- N/A
+
+##### Integration tests
+
+- N/A
+
+##### e2e tests
+
+- N/A
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature gate `RemoveRunonce` is introduced, is disabled by default. Enabling this feature gate will fail the kubelet on startup with runonce mode enable.
```

Comment:
Name the gate after the feature, not the change you intend to make. For example, call it `KubeletRunOnceMode`.